File: frontend/src/api/tsdFiddler.ts
/**
 * TSD Fiddler — API client.
 *
 * Fiddler-verified START/END pairing with machine transactions.
 * Completely separate from old industream-tsd (Presunout flow).
 */

import { apiClient } from './client'

export interface TsdFiddlerStartRequest {
  job: string
  suffix?: string
  oper_num: string
  item?: string
  whse?: string
  kapacity_guid?: string
}

export interface TsdFiddlerEndWorkRequest extends TsdFiddlerStartRequest {
  qty_complete?: number
  qty_scrapped?: number
  oper_complete?: boolean
}

export interface TsdFiddlerResponse {
  status: string
  message: string
  stroj?: string | null
  wc?: string | null
  sp_results?: Record<string, unknown> | null
}

const BASE = '/tsd-fiddler'

/** Start Setup — Infor WrapperSp(TT=1) */
export async function startSetup(req: TsdFiddlerStartRequest): Promise<TsdFiddlerResponse> {
  const { data } = await apiClient.post<TsdFiddlerResponse>(`${BASE}/start-setup`, req)
  return data
}

/** End Setup + auto-Start Work — Infor InsWrapper(TT=2) + Kapacity + ValidateMachine + Wrapper(TT=3) + Mchtrx */
export async function endSetup(req: TsdFiddlerStartRequest): Promise<TsdFiddlerResponse> {
  const { data } = await apiClient.post<TsdFiddlerResponse>(`${BASE}/end-setup`, req)
  return data
}

/** Start Work — Infor ValidateMachine + Wrapper(TT=3) + Kapacity + Mchtrx(H) */
export async function startWork(req: TsdFiddlerStartRequest): Promise<TsdFiddlerResponse> {
  const { data } = await apiClient.post<TsdFiddlerResponse>(`${BASE}/start-work`, req)
  return data
}

/** End Work — Infor InsWrapper(TT=4) + Kapacity + MachineVal + DcSfcMchtrx(J) */
export async function endWork(req: TsdFiddlerEndWorkRequest): Promise<TsdFiddlerResponse> {
  const { data } = await apiClient.post<TsdFiddlerResponse>(`${BASE}/end-work`, req)
  return data
}

File: app/schemas/tsd_fiddler.py
"""TSD Fiddler — Request/Response schemas.

Fiddler-verified START/END flow with machine transactions.
Completely separate from old Presunout-based industream_tsd schemas.
"""

from typing import Optional

from pydantic import BaseModel, Field


class TsdFiddlerStartRequest(BaseModel):
    """Request for start-setup, start-work."""

    job: str = Field(..., min_length=1, max_length=30)
    suffix: str = Field(default="0", max_length=5)
    oper_num: str = Field(..., min_length=1, max_length=10)
    item: str = Field(default="", max_length=30, description="DerJobItem")
    whse: str = Field(default="MAIN", max_length=10)
    kapacity_guid: str = Field(default="0", max_length=50)


class TsdFiddlerEndWorkRequest(TsdFiddlerStartRequest):
    """Request for end-work — includes quantities + flags."""

    qty_complete: float = Field(default=0.0, ge=0)
    qty_scrapped: float = Field(default=0.0, ge=0)
    oper_complete: bool = False


class TsdFiddlerResponse(BaseModel):
    """Response from any TSD Fiddler operation."""

    status: str  # "ok" or "error"
    message: str = ""
    stroj: Optional[str] = None
    wc: Optional[str] = None
    sp_results: Optional[dict] = None  # debug info
